Fix some issues with the Sockets perf tests.

pgavlin opened the pull request:
Operation failure was not properly handled, resulting in test hangs. This also adjusts the ports for the perf tests to avoid conflicts with other tests.

pgavlin committed:
Fix some issues with the Sockets perf tests.

Operation failure was not properly handled, resulting in test hangs.

pgavlin commented:
@CIPop @davidsh @stephentoub PTAL

stephentoub commented on src/Common/tests/System.Net/Sockets/Performance/SocketTestClientAPM.cs:
```diff
@@ -51,12 +60,23 @@ public override void Receive(Action<int> onReceiveCallback)
-            Action<int> callback = (Action<int>)result.AsyncState;
-            int recvBytes = _s.EndReceive(result);
-            callback(recvBytes);
+            Action<int, SocketError> callback = (Action<int, SocketError>)result.AsyncState;
```
Do we need to be concerned about code that receives one of these IAsyncResults doing a similar cast and then invoking the callback?  Are these IAsyncResults exposed out to other code?  Or does the callback protect against that kind of misuse?  Not something to hold up this PR, just something to think through.

stephentoub replied:
Nevermind, I just noticed this is just in the tests.

stephentoub commented:
LGTM

pgavlin committed:
Allow limited retry of failed accept() calls in SocketTestServerAsync.

Once the retry limit is exceeded, the test is failed. This is intended to
prevent test hangs.

pgavlin commented:
Ping. I'd like to get this checked in today, as it's hampering our ability to run tests in the lab in the face of bugs that would otherwise hang the test run.

davidsh commented:
LGTM

CIPop commented on src/Common/tests/System.Net/Sockets/SocketTestServerAsync.cs:
```diff
@@ -168,23 +172,39 @@ private void ProcessAccept(SocketAsyncEventArgs e)
-            // As soon as the client is connected, post a receive to the connection 
-            bool willRaiseEvent = e.AcceptSocket.ReceiveAsync(readEventArgs);
-            if (!willRaiseEvent)
+                Task.Delay(500).Wait();
```
@pgavlin This is an interesting failure: I wouldn't expect accept to fail since there should be connections queued up already.
Did you find why accept fails in general? (Maybe a comment would be useful to debug this code in the future.)

pgavlin replied:
The error I was seeing on OS X was "too many open sockets," so I suspect that the per-process limit on the number of sockets is rather low on the same. Delaying and retrying gives the test just enough time for some of those sockets to close, and the test can continue. It might be better to find a more sophisticated solution in the future.

CIPop commented on src/Common/tests/System.Net/Sockets/Performance/SocketTestClient.cs:
```diff
@@ -118,12 +126,20 @@ private void OnConnect()
-            _log.WriteLine(this.GetHashCode() + " OnSend({0})", bytesSent);
+            _log.WriteLine(this.GetHashCode() + " OnSend({0}, {1})", bytesSent, error);
+
+            // TODO: an error should fail the test.
```
This is debatable in a perf/stress test. We could count failures and depending on resource pressure we could determine the number of acceptable failures for send/receive.
We can use native socket perf tests as a baseline.

pgavlin replied:
Yes, that's a good idea.

CIPop commented on src/Common/tests/System.Net/Sockets/Performance/SocketTestClient.cs:
```diff
@@ -151,14 +167,22 @@ private void OnSendMessage()
+            _log.WriteLine(this.GetHashCode() + " OnSend({0}, {1})", receivedBytes, error);
             _recvBufferIndex += receivedBytes;
 
+            // TODO: an error should fail the test.
```
See above.

CIPop commented:
A few questions regarding retry addition that could signal that existing patterns might not work as expected on other platforms.

LGTM otherwise.
